src/lib/net/src/packets/outgoing/chunk_and_light_data.rs

```diff
@@ -0,0 +1,183 @@
+use crate::errors::NetError;
+use ferrumc_macros::{packet, NetEncode};
+use ferrumc_net_codec::net_types::bitset::BitSet;
+use ferrumc_net_codec::net_types::length_prefixed_vec::LengthPrefixedVec;
+use ferrumc_net_codec::net_types::var_int::VarInt;
+use ferrumc_world::chunk_format::{Chunk, Heightmaps};
+use std::io::{Cursor, Write};
+use std::ops::Not;
+use tokio::io::AsyncWriteExt;
+use tracing::warn;
+
+const SECTIONS: usize = 24; // Number of sections, adjust for your Y range (-64 to 319)
+
+#[derive(NetEncode)]
+pub struct BlockEntity {
+    pub xz: u8,
+    pub y: u16,
+    pub entity_type: VarInt,
+    pub nbt: Vec<u8>,
+}
+
+#[derive(NetEncode)]
+#[packet(packet_id = 0x27)]
+pub struct ChunkAndLightData {
+    pub chunk_x: i32,
+    pub chunk_z: i32,
+    // The binary nbt data
+    pub heightmaps: Vec<u8>,
+    pub data: LengthPrefixedVec<u8>,
+    pub block_entities: LengthPrefixedVec<BlockEntity>,
+    pub sky_light_mask: BitSet,
+    pub block_light_mask: BitSet,
+    pub empty_sky_light_mask: BitSet,
+    pub empty_block_light_mask: BitSet,
+    pub sky_light_arrays: LengthPrefixedVec<LengthPrefixedVec<u8>>,
+    pub block_light_arrays: LengthPrefixedVec<LengthPrefixedVec<u8>>,
+}
+
+impl ChunkAndLightData {
+    pub async fn empty(chunk_x: i32, chunk_z: i32) -> Self {
+        let sky_light_arrays = (0..SECTIONS)
+            .map(|_| LengthPrefixedVec::new(vec![0; 2048]))
+            .collect();
+        let block_light_arrays = (0..SECTIONS)
+            .map(|_| LengthPrefixedVec::new(vec![0; 2048]))
+            .collect();
+        let mut empty_sky_light_mask = BitSet::new(SECTIONS + 2);
+        empty_sky_light_mask.set_all(false);
+        let mut empty_block_light_mask = BitSet::new(SECTIONS + 2);
+        empty_block_light_mask.set_all(false);
+        ChunkAndLightData {
+            chunk_x,
+            chunk_z,
+            heightmaps: Heightmaps::new().serialize_as_network(),
+            data: LengthPrefixedVec::new(vec![0; SECTIONS * 10]),
+            block_entities: LengthPrefixedVec::new(Vec::new()),
+            sky_light_mask: BitSet::new(SECTIONS),
+            block_light_mask: BitSet::new(SECTIONS),
+            empty_sky_light_mask,
+            empty_block_light_mask,
+            sky_light_arrays: LengthPrefixedVec::new(sky_light_arrays),
+            block_light_arrays: LengthPrefixedVec::new(block_light_arrays),
+        }
+    }
```

Comment:
As I said, AsyncWriter lets me write specific data types, whereas the sync versions don't. And right now const optimizations would be a little faster, but once the lighting is generated dynamically we won't be able to use const stuff as much, so i left it this way to avoid restructuring later down the line.